peer/connector.go

```diff
@@ -79,56 +83,100 @@ func (conn *connector) initialise(privateKey []byte, publicKey []byte, connect [
+		go func(conn *connector, c Connection, i int, wg *sync.WaitGroup, ch chan error) {
 
-		client, err := upstream.New(privateKey, publicKey, connectorTimeout)
-		if nil != err {
-			log.Errorf("client[%d]=%q  error: %s", i, address, err)
-			errX = err
-			goto fail
-		}
+			// error function call
+			errF := func(wg *sync.WaitGroup, ch chan error, e error) {
+				ch <- e
+				wg.Done()
+			}
 
-		conn.staticClients[i] = client
-		globalData.connectorClients = append(globalData.connectorClients, client)
+			// for canonicaling the error
+			canonicalErrF := func(c Connection, e error) error {
+				return fmt.Errorf(fmt.Sprintf("client: %q error: %s", c.Address, e))
+			}
 
-		err = client.Connect(address, serverPublicKey)
-		if nil != err {
-			log.Errorf("connect[%d]=%q  error: %s", i, address, err)
-			errX = err
-			goto fail
-		}
-		log.Infof("public key: %x  at: %q", serverPublicKey, c.Address)
+			address, err := util.NewConnection(c.Address)
+			if nil != err {
+				log.Errorf("client[%d]=address: %q  error: %s", i, c.Address, err)
+				errF(wg, ch, canonicalErrF(c, err))
+				return
+			}
+			serverPublicKey, err := hex.DecodeString(c.PublicKey)
+			if nil != err {
+				log.Errorf("client[%d]=public: %q  error: %s", i, c.PublicKey, err)
+				errF(wg, ch, canonicalErrF(c, err))
+				return
+			}
+
+			// prevent connection to self
+			if bytes.Equal(publicKey, serverPublicKey) {
+				err := fault.ErrConnectingToSelfForbidden
+				log.Errorf("client[%d]=public: %q  error: %s", i, c.PublicKey, err)
+				errF(wg, ch, canonicalErrF(c, err))
+				return
+			}
+
+			client, err := upstream.New(privateKey, publicKey, connectorTimeout)
+			if nil != err {
+				log.Errorf("client[%d]=%q  error: %s", i, address, err)
+				errF(wg, ch, canonicalErrF(c, err))
+				return
+			}
+
+			conn.Lock()
+			conn.staticClients[i] = client
+			globalData.connectorClients = append(globalData.connectorClients, client)
+			conn.Unlock()
+
+			err = client.Connect(address, serverPublicKey)
+			if nil != err {
+				log.Errorf("connect[%d]=%q  error: %s", i, address, err)
+				errF(wg, ch, canonicalErrF(c, err))
+				return
+			}
+			log.Infof("public key: %x  at: %q", serverPublicKey, c.Address)
+			wg.Done()
+
+		}(conn, c, i, &wg, errCh)
+	}
+
+	conn.log.Debug("waiting for all static connections...")
+	wg.Wait()
+
+	// drop error channel for getting all errors
+	errs := make([]error, 0)
+	for len(errCh) > 0 {
+		errs = append(errs, <-errCh)
+	}
+
+	// error code for goto fail
+	err := error(nil)
+
+	if len(errs) == 1 {
+		err = errs[0]
+		goto fail
+	} else if len(errs) > 1 {
+		err = compositeError(errs)
+		goto fail
 	}
 
 	// just create sockets for dynamic clients
-	for i := 0; i < maximumDynamicClients; i += 1 {
-		client, err := upstream.New(privateKey, publicKey, connectorTimeout)
-		if nil != err {
+	for i := 0; i < maximumDynamicClients; i++ {
+		client, e := upstream.New(privateKey, publicKey, connectorTimeout)
+		if nil != e {
```

Comment:
`err` is changed to `e`, is line 178 forget to change variable?

Reply:
Yes. I see. My mistake